File: part03/solutions/python/exercise2_distinct_visits_job.py
from pyspark.sql import SparkSession, functions
from pyspark.sql.types import StructType, StructField, StringType, TimestampType

from config import get_checkpoint_location, kafka_input_topic, kafka_deduped_topic

spark = SparkSession.builder.master("local[*]") \
    .config('spark.jars.packages', 'org.apache.spark:spark-sql-kafka-0-10_2.12:3.5.0') \
    .getOrCreate()

input_data = spark.readStream \
    .format("kafka") \
    .option("kafka.bootstrap.servers", "localhost:9094") \
    .option("subscribe", kafka_input_topic()) \
    .option("startingOffsets", "EARLIEST") \
    .load()

event_schema_for_deduplication = StructType([
    StructField("event_id", StringType()),
    StructField("event_time", TimestampType())
])

query = input_data.selectExpr("CAST(value AS STRING)") \
    .select(functions.from_json("value", event_schema_for_deduplication).alias("value_struct"), "value") \
    .select("value_struct.event_time", "value_struct.event_id", "value") \
    .withWatermark("event_time", "10 minutes") \
    .dropDuplicates(["event_id", "event_time"])\
    .drop("event_time", "event_id") # keep only the value which is required for the sink!

writer = query.writeStream.option("checkpointLocation", get_checkpoint_location('deduplication'))\
    .format("kafka")\
    .option("kafka.bootstrap.servers", "localhost:9094")\
    .option("topic", kafka_deduped_topic())

writer.start().awaitTermination()

File: part03/solutions/scala/src/main/scala/com/waitingforcode/Exercise2DistinctVisitsJob.scala
package com.waitingforcode

import org.apache.spark.sql.{SparkSession, functions}
import org.apache.spark.sql.types.{StringType, StructField, StructType, TimestampType}

object Exercise2DistinctVisitsJob {

  def main(args: Array[String]): Unit = {
    val sparkSession = SparkSession.builder().master("local[*]")
      .getOrCreate()

    val inputDataStream = sparkSession.readStream
      .options(Map(
        "kafka.bootstrap.servers" -> "localhost:9094",
        "subscribe" -> InputDataTopicName,
        "startingOffsets" -> "EARLIEST",
      ))
      .format("kafka").load()

    val eventSchema = StructType(Seq(
      StructField("event_id", StringType),
      StructField("event_time", TimestampType)
    ))

    import sparkSession.implicits._
    val query = inputDataStream.selectExpr("CAST(value AS STRING)")
      .select(functions.from_json($"value", eventSchema).alias("value_struct"), $"value")
      .select("value_struct.event_time", "value_struct.event_id", "value")
      .withWatermark("event_time", "10 minutes")
      .dropDuplicates(Seq("event_id", "event_time"))
      // keep only the value which is required for the sink !
      .drop("event_time", "event_id")

    val writer = query.writeStream.option("checkpointLocation", "/tmp/wfc/workshop/part03/checkpoint/deduplication")
      .format("kafka")
      .option("kafka.bootstrap.servers", "localhost:9094")
      .option("topic", DedupedDataTopicName)

    writer.start().awaitTermination()
  }

}
